### PR TITLE
[SPARK-9390][SQL] create a wrapper for array type

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala [hunk 1]
```diff
@@ -0,0 +1,127 @@
+package org.apache.spark.sql.types
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecializedGetters
+import org.apache.spark.unsafe.types.{UTF8String, Interval}
+
+abstract class ArrayData extends SpecializedGetters {
+  def array(): Array[Any]
+
+  def get(ordinal: Int): Any
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala]: the generic `get` and `array` method is useful in interpreted mode, but it's a lot painful to handle `Any` between scala and java, so I put this interface in scala.

**Comment by cloud-fan** [comment]
cc @rxin

**Comment by cloud-fan** [comment]
cc @rxin, I underestimated this work, I took me a lot of time to cover all corn cases...

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 2]
```diff
@@ -683,13 +684,10 @@ class CastSuite extends SparkFunSuite with ExpressionEvalHelper {
...
-        Seq(UTF8String.fromString("123"), UTF8String.fromString("abc"), UTF8String.fromString("")),
-        Map(
-          UTF8String.fromString("a") -> UTF8String.fromString("123"),
-          UTF8String.fromString("b") -> UTF8String.fromString("abc"),
-          UTF8String.fromString("c") -> UTF8String.fromString("")),
-        InternalRow(0)),
+      Row(
+        Seq("123", "abc", ""),
+        Map("a" ->"123", "b" -> "abc", "c" -> ""),
+        Row(0)),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: I think we should avoid using `InternalRow` in test directly, which means we are manually converting external type to internal type. Using `Row` and let converter do this work seems better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.types
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.SpecializedGetters
+import org.apache.spark.unsafe.types.{UTF8String, Interval}
+
+abstract class ArrayData extends SpecializedGetters with Serializable {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala]: can you move this into its own file

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 2 above)
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: That makes sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  //todo: remove this after we handle all types.(map type need special getter)
+  def get(ordinal: Int): Any
+
+  def numElements(): Int
+
+  // todo: need a more efficient way to iterate array type.
+  def toArray(): Array[Any] = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/GenericArrayData.scala]: it seems to me toArray should take a data type? and we should have specialized version for this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/ArrayData.scala [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import org.apache.spark.sql.catalyst.expressions.SpecializedGetters
+
+abstract class ArrayData extends SpecializedGetters with Serializable {
+  // todo: remove this after we handle all types.(map type need special getter)
+  def get(ordinal: Int): Any
+
+  def numElements(): Int
+
+  // todo: need a more efficient way to iterate array type.
+  def toArray(): Array[Any] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/ArrayData.scala]: I think we need specialized version of `toArray`, and get rid of calling generic getter inside it. We can do it after adding special getter for map and removing generic getter.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/ArrayData.scala]: +1

**Comment by cloud-fan** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/json/JacksonParser.scala [hunk 6]
```diff
@@ -165,16 +170,16 @@ private[sql] object JacksonParser {
     builder.result()
   }
 
-  private def convertList(
+  private def convertArray(
       factory: JsonFactory,
       parser: JsonParser,
-      schema: DataType): Seq[Any] = {
-    val builder = Seq.newBuilder[Any]
+      schema: DataType): ArrayData = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JacksonParser.scala]: is this data type of the element? if yes, rename this elementDataType

**Comment by cloud-fan** [comment]
cc @rxin I think it's close, there is still a pyspark doc string test failue:
```
File "/home/jenkins/workspace/SparkPullRequestBuilder/python/pyspark/ml/recommendation.py", line 73, in __main__.ALS
Failed example:
    model.userFactors.orderBy("id").collect()
Expected:
    [Row(id=0, features=[...]), Row(id=1, ...), Row(id=2, ...)]
Got:
    [Row(id=0, features=(None, None, None, None, None, None, None, None, None, None)), Row(id=1, features=(None, None, None, None, None, None, None, None, None, None)), Row(id=2, features=(None, None, None, None, None, None, None, None, None, None))]
```
Does any body have an idea on it?

One possible reason I can think about is: before ArrayData, we use `ArraySeq` as external array value, but now we use `WrappedArray`, will this make difference?

**Comment by rxin** [comment]
cc @davies can you take a look?

**Comment by rxin** [comment]
Maybe this is the problem? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/pythonUDFs.scala#L125

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
cc @rxin, I'll specialize `toArray` when working on MapData.

**Comment by rxin** [comment]
Thanks - I've merged this.

**Pull request closed without merging**